platform/ext/target/nordic_nrf/common/spm_hal.c

```diff
@@ -321,3 +321,68 @@ enum tfm_plat_err_t tfm_spm_hal_nvic_interrupt_enable(void)
 {
     return nvic_interrupt_enable();
 }
+
+bool tfm_spm_hal_has_access_to_region(const void *p, size_t s,
+                                              int flags)
+{
+    cmse_address_info_t tt_base = cmse_TT((void *)p);
+    cmse_address_info_t tt_last = cmse_TT((void *)((uint32_t)p + s - 1));
+
+    uint32_t base_spu_id = tt_base.flags.idau_region;
+    uint32_t last_spu_id = tt_last.flags.idau_region;
+
+    size_t size;
+    uint32_t p_start = (uint32_t)p;
+    int i;
+
+    if ((base_spu_id >= spu_regions_flash_get_start_id()) &&
+        (last_spu_id <= spu_regions_flash_get_last_id())) {
+
+        size = spu_regions_flash_get_last_address_in_region(base_spu_id) + 1 - p_start;
+
+        if (cmse_check_address_range((void *)p_start, size, flags) == 0) {
+            return false;
+        }
+
+        for (i = base_spu_id + 1; i < last_spu_id; i++) {
+            p_start = spu_regions_flash_get_base_address_in_region(i);
+            if (cmse_check_address_range((void *)p_start,
+                spu_regions_flash_get_region_size(), flags) == 0) {
+                return false;
+            }
+        }
+
+        p_start = spu_regions_flash_get_base_address_in_region(last_spu_id);
+        size = (uint32_t)p + s - p_start;
+        if (cmse_check_address_range((void *)p_start, size, flags) == 0) {
+            return false;
+        }
+
+
+    } else if ((base_spu_id >= spu_regions_sram_get_start_id()) &&
+        (last_spu_id <= spu_regions_sram_get_last_id())) {
+
+        size = spu_regions_sram_get_last_address_in_region(base_spu_id) + 1 - p_start;
+        if (cmse_check_address_range((void *)p_start, size, flags) == 0) {
+            return false;
+        }
+
+        for (i = base_spu_id + 1; i < last_spu_id; i++) {
+            p_start = spu_regions_sram_get_base_address_in_region(i);
+            if (cmse_check_address_range((void *)p_start,
+                spu_regions_sram_get_region_size(), flags) == 0) {
+                return false;
+            }
+        }
+
+        p_start = spu_regions_sram_get_base_address_in_region(last_spu_id);
+        size = (uint32_t)p + s - p_start;
+        if (cmse_check_address_range((void *)p_start, size, flags) == 0) {
+            return false;
+        }
+    } else {
+        return false;
+    }
```

Comment:
Unless I'm missing something, I feel this could be acheived a lot simpler, something like
```suggestion
    for (i = base_spu_id; i <= last_spu_id; i++) {
        uint32_t start = max(spu_get_base_address_in_region(i), p_start);
        uint32_t end = min(spu_get_last_address_in_region(i) + 1, p_start + s);
        if (cmse_check_address_range((void *)start, end - start, flags) == 0) {
            return false;
        }
    }
```
You'd have to merge the `spu_get_base_address_in_region()` etc as I suggested above.

Reply:
I like the idea of the min and max (though, actually, they add conditional check, i.e. some overhead)

Also merging the spu_get_.... functions will require conditional checks inside the functions, to distinguish between flash, ram and invalid regions.
Finally, we cannot just enumerate  on -i- because we need to cover the case of base_spu_id pointing to FLASH and last_spu_id pointing to SRAM. Some extra checks will be needed there.

I think the main point in this function would be to be fast, not have minimal footprint. We are targeting nrf53 and 91 with lots of flash. We care more about processing time.

Reply:
How often will this function be called?